rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java

```diff
@@ -44,33 +61,88 @@
  */
 public class FlowNodeUtil {
 
+    protected static final Cache<String,List<FlowNode>> executionCache = CacheBuilder.newBuilder().maximumSize(100).build();
+
+
+    // Larger cache of run data, for completed runs, keyed by flowexecution url, useful for serving info
+    // Actually can be used to serve Stage data too
+    // Because the RunExt caps the total elements returned, and this is fully realized, this is the fastest way
+    public static final Cache<String, RunExt> runData = CacheBuilder.newBuilder().maximumSize(1000).build();
+
+    static final Cache<FlowNode,String> execNodeNameCache = CacheBuilder.newBuilder().weakKeys().expireAfterAccess(1, TimeUnit.HOURS).build();
+
     private static final Logger LOGGER = Logger.getLogger(FlowNodeUtil.class.getName());
 
     private FlowNodeUtil() {
     }
 
-    public static long getNodeExecDuration(FlowNode node) {
-        List<FlowNode> childNodes = getChildNodes(node);
-        if (!childNodes.isEmpty()) {
-            long startTime = TimingAction.getStartTime(node);
-            long endTime = TimingAction.getStartTime(childNodes.get(childNodes.size() - 1));
+    @CheckForNull
+    public static RunExt getCachedRun(FlowExecution ex) {
+        try {
+            if (ex != null) {
+                return runData.getIfPresent(ex.getUrl());
+            }
+            return null;
+        } catch (IOException ioe) {
+            LOGGER.severe("Can't get execution url for execution, IOException!");
+            throw new IllegalStateException(ioe);
+        }
+    }
 
-            if (startTime == 0) {
-                // The node is running and the time has not been marked on it yet.  Return 0 as the duration for now.
-                return 0;
+    public static void cacheRun(FlowExecution exec, RunExt run) {
+        if (exec != null && runData != null && exec.isComplete()) {
+            try {
+                runData.put(exec.getUrl(), run);
+            } catch (IOException ioe) {
+                LOGGER.severe("Can't get execution url for execution, IOException!");
             }
+        }
+    }
+
+    /** The nuclear option: clear all caches */
+    public static void invalidateAllCaches() {
```

Comment:
:ant: having such a method is an antipattern. Define your caches with `Jenkins` scope to begin with and this is not necessary.